File: lib/scanner/plugins/servers/nginx.rb
# frozen_string_literal: true

module Yawast
  module Scanner
    module Plugins
      module Servers
        class Nginx < Yawast::Scanner::Base
          def self.check_banner(banner)
            Yawast::Shared::Output.log_hash 'vulnerabilities',
                                            'nginx_version_exposed',
                                            {vulnerable: false, version: nil}

            # don't bother if this doesn't include nginx
            return unless banner.include? 'nginx/'

            Yawast::Utilities.puts_warn "nginx Version: #{banner}"
            puts ''

            Yawast::Shared::Output.log_hash 'vulnerabilities',
                                            'nginx_version_exposed',
                                            {vulnerable: true, version: banner}
          end

          def self.check_all(uri)
            check_status_page uri.copy
          end

          def self.check_status_page(uri)
            uri.path = '/status'
            uri.query = '' unless uri.query.nil?

            body = Yawast::Shared::Http.get(uri)

            if body.include? 'Active connections:'
              Yawast::Utilities.puts_vuln "Nginx status page found: #{uri}"

              Yawast::Shared::Output.log_hash 'vulnerabilities',
                                              'nginx_status_found',
                                              {vulnerable: true, uri: uri, body: body}

              puts ''
            else
              Yawast::Shared::Output.log_hash 'vulnerabilities',
                                              'nginx_status_found',
                                              {vulnerable: false, uri: uri, body: body}
            end
          end
        end
      end
    end
  end
end

File: yawast/scanner/plugins/http/servers/nginx.py
from typing import List, cast
from urllib.parse import urljoin

from packaging import version

from yawast.scanner.plugins.http import version_checker, response_scanner
from yawast.reporting.enums import Vulnerabilities
from yawast.scanner.plugins.result import Result
from yawast.shared import network


def check_all(url: str) -> List[Result]:
    results = []

    results += check_status(url)

    return results


def check_banner(banner: str, raw: str, url: str) -> List[Result]:
    if not banner.startswith("nginx"):
        return []

    results = []

    if "/" in banner:
        # we've got a Nginx version
        results.append(
            Result(
                f"Nginx Version Exposed: {banner}",
                Vulnerabilities.HTTP_BANNER_NGINX_VERSION,
                url,
                raw,
            )
        )

        # parse the version, and get the latest version - see if the server is up to date
        ver = cast(version.Version, version.parse(banner.split("/")[1]))
        curr_version = version_checker.get_latest_version("nginx", ver)

        if curr_version is not None and curr_version > ver:
            results.append(
                Result(
                    f"Nginx Outdated: {ver} - Current: {curr_version}",
                    Vulnerabilities.SERVER_NGINX_OUTDATED,
                    url,
                    raw,
                )
            )
    else:
        # this means that it's just a generic banner, with no info
        results.append(
            Result(
                "Generic Nginx Server Banner Found",
                Vulnerabilities.HTTP_BANNER_GENERIC_NGINX,
                url,
                raw,
            )
        )

    return results


def check_status(url: str) -> List[Result]:
    results = []
    search = ["status/", "stats/"]

    for path in search:
        target = urljoin(url, path)

        res = network.http_get(target, False)
        body = res.text

        if res.status_code == 200 and "Active connections:" in body:
            results.append(
                Result(
                    f"Nginx status page found: {target}",
                    Vulnerabilities.SERVER_NGINX_STATUS_EXPOSED,
                    target,
                    [
                        network.http_build_raw_request(res.request),
                        network.http_build_raw_response(res),
                    ],
                )
            )

        results += response_scanner.check_response(target, res)

    return results
